### PR TITLE
TMA-868: deprecated flag now propagates the value to the replacement …

### PR DESCRIPTION
**Pull request by Hahihula**
…param

### CODE REVIEW

File: spec/unit/helpers/check_helper_spec.rb [hunk 1]
```diff
@@ -50,6 +50,43 @@
...
+    end
+
+    let(:other_params) do
+      raw_params = { 'update_preference' => { 'keep_data' => false,
+                                              'cascade_drops' => true } }
+      GoodData::LCM2.convert_to_smart_hash(raw_params)
+    end
+
+    let(:spec) do
+      GoodData::LCM2::BaseAction.define_params(self) do
```
Comment (yenkins) [inline spec/unit/helpers/check_helper_spec.rb]: Similar code found in check_helper_spec.rb:41, check_helper_spec.rb:67 (mass = 38)

File: spec/unit/helpers/check_helper_spec.rb [hunk 2]
```diff
@@ -50,6 +50,43 @@
     expect { params[:test_param_three] }.to raise_error(/not defined in the specification/)
   end
 
+  context 'when created from stringified hash' do
+    let(:params) do
```
Comment (yenkins) [inline spec/unit/helpers/check_helper_spec.rb]: Similar code found in check_helper_spec.rb:54, check_helper_spec.rb:60 (mass = 32)

File: spec/unit/helpers/check_helper_spec.rb [hunk 3]
```diff
@@ -50,6 +50,43 @@
...
+      raw_params = { 'update_preference' => { 'keep_data' => false,
+                                              'allow_cascade_drops' => true } }
+      GoodData::LCM2.convert_to_smart_hash(raw_params)
+    end
+
+    let(:other_params) do
```
Comment (yenkins) [inline spec/unit/helpers/check_helper_spec.rb]: Similar code found in check_helper_spec.rb:54, check_helper_spec.rb:60 (mass = 32)

File: lib/gooddata/lcm/helpers/check_helper.rb [hunk 4]
```diff
@@ -27,6 +27,7 @@ def check_params(specification, params)
               end
 
               if specification[param_name][:opts][:deprecated]
+                params[specification[param_name][:opts][:replacement]] = value
```
Comment (yenkins) [inline lib/gooddata/lcm/helpers/check_helper.rb]: Calls 'specification[param_name]' 7 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/helpers/check_helper.rb]: Calls 'specification[param_name][:opts]' 6 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/helpers/check_helper.rb]: Calls 'specification[param_name][:opts][:replacement]' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/helpers/check_helper.rb]: Refers to 'specification' more than self (maybe move it to another class?) (FeatureEnvy)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/923/ : SUCCESS in 26s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/774/ : FAILURE in 36s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1042/ : SUCCESS in 4m 23s

File: spec/unit/helpers/check_helper_spec.rb [hunk 5]
```diff
@@ -50,6 +50,42 @@
...
+      GoodData::LCM2.convert_to_smart_hash(raw_params)
+    end
+
+    let(:other) do
+      raw_params = { 'update_preference' => { 'keep_data' => false, 'cascade_drops' => true } }
+      GoodData::LCM2.convert_to_smart_hash(raw_params)
+    end
+
+    let(:spec) do
+      GoodData::LCM2::BaseAction.define_params(self) do
```
Comment (yenkins) [inline spec/unit/helpers/check_helper_spec.rb]: Similar code found in check_helper_spec.rb:41, check_helper_spec.rb:66 (mass = 38)

File: spec/unit/helpers/check_helper_spec.rb [hunk 6]
```diff
@@ -50,6 +50,42 @@
...
   end
 
+  context 'when created from stringified hash' do
+    let(:params) do
+      raw_params = { 'update_preference' => { 'keep_data' => false,
+                                              'allow_cascade_drops' => true } }
+      GoodData::LCM2.convert_to_smart_hash(raw_params)
+    end
+
+    let(:other) do
```
Comment (yenkins) [inline spec/unit/helpers/check_helper_spec.rb]: Similar code found in check_helper_spec.rb:54, check_helper_spec.rb:60 (mass = 32)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/926/ : SUCCESS in 14s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/780/ : FAILURE in 34s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1048/ : SUCCESS in 35s

File: spec/unit/helpers/check_helper_spec.rb [hunk 7]
```diff
@@ -50,6 +50,46 @@
...
+<<<<<<< HEAD
+      raw_params = { 'update_preference' => { 'cascade_drops' => true, 'keep_data' => false } }
+=======
+      raw_params = { 'update_preference' => { 'keep_data' => false, 'cascade_drops' => true } }
+>>>>>>> fa6204dc515ab459195c4e0e733061b28d2505b4
+      GoodData::LCM2.convert_to_smart_hash(raw_params)
+    end
```
Comment (yenkins) [inline spec/unit/helpers/check_helper_spec.rb]: Has unexpected token kend (Syntax)

File: spec/unit/helpers/check_helper_spec.rb [hunk 8]
```diff
@@ -50,6 +50,46 @@
...
+      GoodData::LCM2::Helpers.check_params(spec, other)
+      expect(other[:update_preference][:keep_data]).to be(false)
+      expect(other[:update_preference][:cascade_drops]).to be(true)
+    end
+
+    it 'it propagates the values of deprecated params to the replacement' do
+      GoodData::LCM2::Helpers.check_params(spec, other)
+      expect(other[:update_preference][:allow_cascade_drops]).to be(true)
+    end
+  end
```
Comment (yenkins) [inline spec/unit/helpers/check_helper_spec.rb]: Has unexpected token kend (Syntax)

File: spec/unit/helpers/check_helper_spec.rb [hunk 9]
```diff
@@ -50,6 +50,46 @@
...
+    let(:params) do
+      raw_params = { 'update_preference' => { 'keep_data' => false,
+                                              'allow_cascade_drops' => true } }
+      GoodData::LCM2.convert_to_smart_hash(raw_params)
+    end
+
+    let(:other) do
+<<<<<<< HEAD
+      raw_params = { 'update_preference' => { 'cascade_drops' => true, 'keep_data' => false } }
+=======
```
Comment (yenkins) [inline spec/unit/helpers/check_helper_spec.rb]: Has unexpected token teqq (Syntax)

Comment (yenkins) [inline spec/unit/helpers/check_helper_spec.rb]: unexpected token tEQQ
(Using Ruby 2.3 parser; configure using `TargetRubyVersion` parameter, under `AllCops`)

File: spec/unit/helpers/check_helper_spec.rb [hunk 10]
```diff
@@ -50,6 +50,46 @@
...
 
+  context 'when created from stringified hash' do
+    let(:params) do
+      raw_params = { 'update_preference' => { 'keep_data' => false,
+                                              'allow_cascade_drops' => true } }
+      GoodData::LCM2.convert_to_smart_hash(raw_params)
+    end
+
+    let(:other) do
+<<<<<<< HEAD
```
Comment (yenkins) [inline spec/unit/helpers/check_helper_spec.rb]: Has unexpected token tlshft (Syntax)

Comment (yenkins) [inline spec/unit/helpers/check_helper_spec.rb]: unexpected token tLSHFT
(Using Ruby 2.3 parser; configure using `TargetRubyVersion` parameter, under `AllCops`)

File: spec/unit/helpers/check_helper_spec.rb [hunk 11]
```diff
@@ -50,6 +50,46 @@
...
+      raw_params = { 'update_preference' => { 'cascade_drops' => true, 'keep_data' => false } }
+=======
+      raw_params = { 'update_preference' => { 'keep_data' => false, 'cascade_drops' => true } }
+>>>>>>> fa6204dc515ab459195c4e0e733061b28d2505b4
```
Comment (yenkins) [inline spec/unit/helpers/check_helper_spec.rb]: Has unexpected token trshft (Syntax)

Comment (yenkins) [inline spec/unit/helpers/check_helper_spec.rb]: unexpected token tRSHFT
(Using Ruby 2.3 parser; configure using `TargetRubyVersion` parameter, under `AllCops`)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/927/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/782/ : FAILURE in 35s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1050/ : FAILURE in 32s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/929/ : SUCCESS in 11s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/784/ : SUCCESS in 31s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1052/ : FAILURE in 34s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/930/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/785/ : SUCCESS in 29s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1053/ : SUCCESS in 32s

**Comment by panjan** [comment]
ok to merge

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-merge-compliance-check/331/ : SUCCESS in 15s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/786/ : SUCCESS in 1m 55s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1054/ : SUCCESS in 1m 58s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1310/ : SUCCESS in 54m 16s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/1073/ : UNSTABLE in 21m 57s

File: lib/gooddata/lcm/helpers/check_helper.rb [hunk 12]
```diff
@@ -27,6 +27,11 @@ def check_params(specification, params)
               end
 
               if specification[param_name][:opts][:deprecated]
+                if specification[specification[param_name][:opts][:replacement]][:type].class == type.class
```
Comment (yenkins) [inline lib/gooddata/lcm/helpers/check_helper.rb]: Calls 'specification[param_name]' 11 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/helpers/check_helper.rb]: Calls 'specification[param_name][:opts]' 10 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/helpers/check_helper.rb]: Calls 'specification[param_name][:opts][:replacement]' 6 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/helpers/check_helper.rb]: Calls 'specification[specification[param_name][:opts][:replacement]]' 3 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/helpers/check_helper.rb]: Calls 'specification[specification[param_name][:opts][:replacement]][:type]' 3 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/helpers/check_helper.rb]: Calls 'type.class' 2 times (DuplicateMethodCall)

File: lib/gooddata/lcm/types/base_type.rb [hunk 13]
```diff
@@ -20,6 +20,14 @@ def check_params(specification, params)
           end
         end
 
+        def migrate?(to_type)
+          migrate(nil, to_type)
+        end
+
+        def migrate(value, to_type)
+          value.nil? ? (false) : (fail "Can not migrate #{self.class} to #{to_type}")
```
Comment (yenkins) [inline lib/gooddata/lcm/types/base_type.rb]: Performs a nil-check (NilCheck)

File: lib/gooddata/lcm/types/scalar/string.rb [hunk 14]
```diff
@@ -16,6 +16,15 @@ class StringType < BaseType
         def check(value)
           value.is_a?(String)
         end
+
+        def migrate(value, to_type)
+          case to_type
+          when GoodData::LCM2::Type::ArrayType
+            value.nil? ? (true) : ([value])
```
Comment (yenkins) [inline lib/gooddata/lcm/types/scalar/string.rb]: Calls 'value.nil?' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/types/scalar/string.rb]: Refers to 'value' more than self (maybe move it to another class?) (FeatureEnvy)

Comment (yenkins) [inline lib/gooddata/lcm/types/scalar/string.rb]: Performs a nil-check (NilCheck)

File: lib/gooddata/lcm/helpers/check_helper.rb [hunk 15]
```diff
@@ -27,6 +27,11 @@ def check_params(specification, params)
...
+                  params[specification[param_name][:opts][:replacement]] = value
+                elsif type.migrate?(specification[specification[param_name][:opts][:replacement]][:type])
+                  params[specification[param_name][:opts][:replacement]] = type.migrate(value,specification[specification[param_name][:opts][:replacement]][:type])
```
Comment (yenkins) [inline lib/gooddata/lcm/helpers/check_helper.rb]: Style/SpaceAfterComma: Space missing after comma.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/933/ : SUCCESS in 12s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/792/ : FAILURE in 2m 24s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1060/ : SUCCESS in 2m 55s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/934/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/793/ : SUCCESS in 2m 04s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1061/ : SUCCESS in 2m 04s

**Comment by Hahihula** [comment]
extended test

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/935/ : SUCCESS in 12s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1314/ : FAILURE in 38m 24s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/1077/ : FAILURE in 53m 39s

**Comment by Hahihula** [comment]
extended test - develop

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/936/ : SUCCESS in 14s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1315/ : FAILURE in 56m 05s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/1078/ : UNSTABLE in 9m 46s

**Comment by Hahihula** [comment]
extended test - staging

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/939/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1321/ : FAILURE in 48m 38s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/1084/ : UNSTABLE in 23m 15s

**Comment by Hahihula** [comment]
extended test

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/941/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1323/ : FAILURE in 22m 00s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/1086/ : UNSTABLE in 5m 07s

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/948/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1325/ : FAILURE in 53m 36s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/1088/ : UNSTABLE in 39m 10s

**Comment by Hahihula** [comment]
extended test

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/2/ : SUCCESS in 13s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1327/ : SUCCESS in 57m 13s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/1090/ : UNSTABLE in 23m 02s

**Comment by Hahihula** [comment]
extended test

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/6/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1331/ : SUCCESS in 55m 46s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/1094/ : UNSTABLE in 21m 47s

**Comment by Hahihula** [comment]
extended test

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/12/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1334/ : SUCCESS in 52m 46s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/1097/ : ABORTED in 12m 46s

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/18/ : SUCCESS in 10s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1339/ : SUCCESS in 54m 07s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/1102/ : UNSTABLE in 22m 43s

File: lib/gooddata/models/project.rb [hunk 16]
```diff
@@ -966,9 +966,18 @@ def token
...
+      if tags.is_a?(Array)
+        if tags.length == 1 && tags.first =~ /,/
+          tags = tags.first.split(',').map(&:strip)
+        end
+      else
+        tags = tags.split(',').map(&:strip)
+      end
 
       objects = tags.map do |tag|
+        p '======================='
```
Comment (yenkins) [inline lib/gooddata/models/project.rb]: Calls 'p '======================='' 2 times (DuplicateMethodCall)

File: lib/gooddata/models/project.rb [hunk 17]
```diff
@@ -966,9 +966,18 @@ def token
     alias_method :member, :get_user
 
     def find_by_tag(tags)
-      tags = tags.split(',').map(&:strip) unless tags.is_a?(Array)
+      if tags.is_a?(Array)
+        if tags.length == 1 && tags.first =~ /,/
```
Comment (yenkins) [inline lib/gooddata/models/project.rb]: Calls 'tags.first' 2 times (DuplicateMethodCall)

File: lib/gooddata/models/project.rb [hunk 18]
```diff
@@ -966,9 +966,18 @@ def token
     alias_method :member, :get_user
 
     def find_by_tag(tags)
-      tags = tags.split(',').map(&:strip) unless tags.is_a?(Array)
+      if tags.is_a?(Array)
```
Comment (yenkins) [inline lib/gooddata/models/project.rb]: Refers to 'tags' more than self (maybe move it to another class?) (FeatureEnvy)

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/21/ : SUCCESS in 11s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/827/ : SUCCESS in 2m 24s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1095/ : SUCCESS in 2m 02s

**Pull request closed without merging**